hierarqcal/core.py

```diff
@@ -47,8 +47,47 @@ def __init__(self, function=None, n_symbols=0, arity=2, symbols=None):
             symbols (list, optional): List of symbol values (rotation angles). Elements can be either sympy symbols, complex numbers, or qiskit parameters. Defaults to None.
         """
         self.function = function
-        self.n_symbols = n_symbols
-        self.arity = arity
+        if isinstance(self.function, str):
+            # automatically calculating arity and n_symbols from string
+            # by getting the number of unique bits and parameters
+            substrings = self.function.split(';')
+            # Remove any leading or trailing whitespaces from each substring
+            substrings = [substring.strip() for substring in substrings]
+            unique_bits = []
+            unique_params = []
+            for substring in substrings:
+                
+                # Isolating the parameters and bits in the substring
+                param_start_index = substring.find('(')
+                param_end_index = substring.find(')')
+
+                bits_start_index = substring.find('^')
+                bits_end_index = len(substring)
+
+                # checking if parameter is unique
+                params_string = substring[param_start_index+1:param_end_index]
+                if params_string == '':
+                    pass
+                else:
+                    param_list = params_string.split(',')
+                    # Remove any leading or trailing whitespaces from each substring
+                    param_list = [param_entry.strip() for param_entry in param_list]
```

Comment:
Minor thing: Can we change this so that param_list only adds param_entries that is not in unique_params. i.e. do the if check within this list comprehension and then extend unique_params on the next line if param_list is not empty. This way we don't have to loop twice through param_list

Reply:
yes, done